Support the function COL_LENGTH()

roshan0708 opened the pull request:
### Description

Implement a function COL_LENGTH(table_name, column_name) which returns the Transact-SQL length of the specified column in bytes based on its data type and length modifiers.

Input-
CREATE TABLE test_table(c1 nvarchar(40), c2 varchar(MAX), c3 char(10));
SELECT COL_LENGTH('test_table', 'c1');
SELECT COL_LENGTH('test_table', 'c2');
SELECT COL_LENGTH('test_table', 'c3');
GO

Output-
80
-1
10

Signed-off-by: Roshan Kanwar [rskanwar@amazon.com](mailto:rskanwar@amazon.com)

### Issues Resolved

BABEL-3489

### Test Scenarios Covered ###
* **Use case based -** Added test cases for columns declared using all data-types supported by BBF


* **Boundary conditions -** Added test cases for handling columns declared with `MAX` specifier


* **Arbitrary inputs -** Added test cases  where inputs can be `NULL`


* **Negative test cases -** Added


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

hash-16 commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
You should add the function in upgrade script file to fix upgrade test failures.

roshan0708 replied:
Should I add the function in `babelfishpg_tsql--3.2.0--3.3.0.sql` or any upgrade file?

hash-16 replied:
Yes under upgrade/babelfishpg_tsql--3.2.0--3.3.0.sql .

shah-nirmit commented on test/JDBC/input/BABEL-3489-vu-verify.sql:
```diff
@@ -0,0 +1,98 @@
+USE babel_3489_test_db;
```
Can we add dependencies on this new function ?
Example View,Procedure referencing it

shalinilohia50 commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -86,6 +86,7 @@ BABEL-3402
 BABEL-3474
 BABEL-3478
 BABEL-3486
+BABEL-3489
```
Lets rename the test file as `col_length`.

shalinilohia50 commented on test/JDBC/input/col_length-vu-verify.sql:
```diff
@@ -0,0 +1,110 @@
+USE babel_3489_test_db;
+GO
+
```
Add some tests with regular `col_length` function call (not inside view, procedures etc).
Include negative test cases too.

shalinilohia50 commented on test/JDBC/input/col_length-vu-prepare.sql:
```diff
@@ -0,0 +1,217 @@
+    col_datetimeoffset DATETIMEOFFSET,
+    col_uniqueidentifier UNIQUEIDENTIFIER
+);
+GO
```
Can you please create a table in another schema and write few tests around that?

Commit:
Support the function COL_LENGTH()

Implement a function COL_LENGTH(table_name, column_name) which returns the
Transact-SQL length of the specified column in bytes based on its data type and length modifiers.

Task: BABEL-3489
Signed-off-by: Roshan Kanwar <rskanwar@amazon.com>

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-3489

roshan0708 committed:
Fix upgrade file changes

Signed-off-by: Roshan Kanwar <rskanwar@amazon.com>

shalinilohia50 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql:
```diff
@@ -44,3 +44,98 @@ DROP PROCEDURE sys.babelfish_drop_deprecated_object(varchar, varchar, varchar);
 
 -- Reset search_path to not affect any subsequent scripts
 SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
```
Please move the function definition before line:45

roshan0708 committed:
Fix upgrade file changes

Signed-off-by: Roshan Kanwar <rskanwar@amazon.com>

hash-16 reviewed (commented):
Please resolve the merge conflicts.

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-3489

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-3489

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-3489

roshan0708 committed:
Fix upgrade file changes

Signed-off-by: Roshan Kanwar <rskanwar@amazon.com>

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-3489

hash-16 commented on test/JDBC/expected/col_length-vu-prepare.out:
```diff
@@ -0,0 +1,230 @@
+-- Invalid table, should return NULL
+CREATE FUNCTION col_length_prepare_f15()
+RETURNS INT AS
+BEGIN
```
You can add few having empty text as parameters , using some cast in parameters and also using this function in some tables and as expressions.

hash-16 commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3926,6 +3926,101 @@ RETURNS SETOF RECORD
+RETURNS SMALLINT AS $BODY$
+    DECLARE
+        col_name TEXT;
+        object_id oid;
```
From which schema are we using text data type from ?

roshan0708 replied:
from `pg_catalog` schema

roshan0708 committed:
Add more test cases for COL_LENGTH() function

Signed-off-by: Roshan Kanwar <rkanwar0708@gmail.com>

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-3489

jsudrik reviewed (approved):
Based on earlier approvals.

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql:
```diff
@@ -302,6 +302,101 @@ ALTER FUNCTION sys.power(IN arg1 TINYINT, IN arg2 NUMERIC) STRICT;
 -- Please have this be one of the last statements executed in this upgrade script.
 DROP PROCEDURE sys.babelfish_drop_deprecated_object(varchar, varchar, varchar);
```
Can you please move this to end of the script?
